R/contribution_matrix.R

```diff
@@ -0,0 +1,588 @@
+      X_star[1:nrow(X), 1:nrow(X)] <- diag(1, nrow = nrow(X))
+      #Populate the bottom left of X_star
+      X_star[(nrow(X) + 1):nrow(X_star), 1:nrow(X)] <- diag(1, nrow = nrow(X))
+      #Populate the bottom right of X_star
+      X_star[(nrow(X) + 1):nrow(X_star), (nrow(X) + 1):ncol(X_star)] <- -X
+      
+      A <- solve(t(X_star) %*% solve(V_star) %*% X_star) %*% t(X_star) %*% solve(V_star)
+      A_bottom_left <- A[(nrow(X) + 1):nrow(A), 1:nrow(X)]
+      
+      if(all_parameters){
+        contribution <- Z %*% A_bottom_left
+      } else if(!all_parameters){
+        contribution <- A_bottom_left
+      }
+      
+    } else if(cov_parameters == "Exchangeable"){
+      
+      if(is.null(std_dev_beta)){stop("Must specify 'std_dev_beta' when cov_parameters == 'Exchangeable'")}
+      
+      Lambda_beta <- CreateLambdaBetaMatrix(treatments = treatments, std_dev_beta = std_dev_beta)
+      
+      #Create V_star with the correct dimensions and 0 everywhere
+      V_star <- matrix(0, nrow = nrow(V) + nrow(Lambda_tau) + nrow(Lambda_beta), ncol = ncol(V) + ncol(Lambda_tau) + ncol(Lambda_beta))
+      rownames(V_star) <- c(rownames(V), rownames(Lambda_tau), rownames(Lambda_beta))
+      colnames(V_star) <- c(colnames(V), colnames(Lambda_tau), colnames(Lambda_beta))
+      #Populate the top left of V_star
+      V_star[1:nrow(V), 1:ncol(V)] <- V
+      #Populate the middle of V_star
+      V_star[(nrow(V) + 1):(nrow(V) + nrow(Lambda_tau)), (ncol(V) + 1):(ncol(V) + ncol(Lambda_tau))] <- Lambda_tau
+      #Populate the bottom right of V_star
+      V_star[(nrow(V) + nrow(Lambda_tau) + 1):nrow(V_star), (ncol(V) + ncol(Lambda_tau) + 1):ncol(V_star)] <- Lambda_beta
+      
+      #The left half of X, corresponding to the treatment effects
+      X_d <- X[, 1:(n_treatments - 1)]
+      #The right half of X, corresponding to the covariate effects
+      X_beta <- X[, n_treatments:(2 * (n_treatments - 1))]
+      
+      #Create X_star with the correct dimensions and 0 everywhere
+      X_star <- matrix(0, nrow = 2 * nrow(X) + ncol(X_d), ncol = nrow(X) + 2 * ncol(X_d) + 1)
+      #Populate the top left of X_star
+      X_star[1:nrow(X), 1:nrow(X)] <- diag(1, nrow = nrow(X))
+      #Populate the middle left of X_star
+      X_star[(nrow(X) + 1):(2 * nrow(X)), 1:nrow(X)] <- diag(1, nrow = nrow(X))
+      #Populate the middle second column of X_star
+      X_star[(nrow(X) + 1):(2 * nrow(X)), (nrow(X) + 1):(nrow(X) + ncol(X_d))] <- -X_d
+      #Populate the middle third column of X_star
+      X_star[(nrow(X) + 1):(2 * nrow(X)), (nrow(X) + ncol(X_d) + 1):(nrow(X) + 2 * ncol(X_d))] <- -X_beta
+      #Populate the bottom third column of X_star
+      X_star[(2 * nrow(X) + 1):nrow(X_star), (nrow(X) + ncol(X_d) + 1):(nrow(X) + 2 * ncol(X_d))] <- diag(1, nrow = ncol(X_d))
+      #Populate the bottom right column of X_star
+      X_star[(2 * nrow(X) + 1):nrow(X_star), (nrow(X) + 2 * ncol(X_d) + 1):ncol(X_star)] <- -matrix(1, nrow = ncol(X_d), ncol = 1)
```

Comment:
Are any of these recalculated indices which could be calculated once and stored in a variable?

Reply:
I have gone further and created an index for all of them, even those that are used only once. It looks neater.